src/Optimizer/Transformer/OptimizeHeroImages.php

```diff
@@ -215,6 +222,17 @@ private function findHeroImages(Document $document, $maxHeroImageCount)
                 $seenParagraphCount++;
             }
 
+            if ($node->tagName === Extension::IMG && ! $node->hasAttribute(Attribute::DATA_HERO)) {
+                $highFetchpriorityImage = $document->xpath->query(
+                    self::NOSCRIPT_IMG_FETCHPRIORITY_HIGH_XPATH_QUERY,
+                    $node
+                )->item(0);
+
+                if ($highFetchpriorityImage instanceof Element) {
+                    $node->setAttribute(Attribute::DATA_HERO, '');
+                }
+            }
+
             $heroImage = $this->detectImageWithAttribute($node, Attribute::DATA_HERO);
             if ($heroImage) {
```

Comment:
Alternatively, this could be done without setting `data-hero`, right?

```suggestion

            $hasFetchpriorityHigh = false;
            if ($node->tagName === Extension::IMG && ! $node->hasAttribute(Attribute::DATA_HERO)) {
                $highFetchpriorityImage = $document->xpath->query(
                    self::NOSCRIPT_IMG_FETCHPRIORITY_HIGH_XPATH_QUERY,
                    $node
                )->item(0);

                if ($highFetchpriorityImage instanceof Element) {
                    $hasFetchpriorityHigh = true;
                }
            }

            $heroImage = $this->detectImageWithAttribute($node, Attribute::DATA_HERO);
            if ($heroImage || $hasFetchpriorityHigh) {
```

If absent, then `generateImg` will be adding it anyway:

https://github.com/ampproject/amp-toolbox-php/blob/89707d0e55a4cbb4be6c1b3a3ed23d9193beb4a6/src/Optimizer/Transformer/OptimizeHeroImages.php#L601-L603

Reply:
But here the node should be transformed to HeroImage. 

https://github.com/ampproject/amp-toolbox-php/blob/89707d0e55a4cbb4be6c1b3a3ed23d9193beb4a6/src/Optimizer/Transformer/OptimizeHeroImages.php#L288

Adding `data-hero` will do that.